Error analysis for solving systems of linear equations

Posterior analysis

Ax = b for the case where:

Augmented matrix:  (A|b) =\begin{bmatrix}  
  0.2161 & 0.1441 &| 0.1440 \\
  1.2969 & 0.8648 &| 0.8642 \end{bmatrix}

  Check an approximate solution: $\tilde{x}= \left[
         	\begin{array}{c}
         	.9911  \\\\
         	-.4870
        	\end{array}
    	\right]   $


$r=b-A\tilde{x}= \left(
         	\begin{array}{c}
         	-10^{-8}  \\\\
         	10^{-8}
        	\end{array}
    	\right) $



In [1]:
import numpy as np
Ap = [[.2161,.1441], [1.2969,.8648]]
Bp=[[0.1440],[0.8642]]
A = np.array(Ap)
print(A)
B = np.array(Bp)
print(B)
Ainv=np.linalg.inv(A)
x=np.matmul(Ainv,B)
print("Exact solution: x= ", x)
xtp=[[0.9911],[-0.487]]
xt=np.array(xtp)
rp=print(B-np.matmul(A,xt))
r=print(B-np.matmul(A,x))

[[0.2161 0.1441]
 [1.2969 0.8648]]
[[0.144 ]
 [0.8642]]
Exact solution: x=  [[ 2.        ]
 [-2.00000001]]
[[-9.99999999e-09]
 [ 1.00000001e-08]]
[[-4.38496184e-11]
 [-2.63158939e-10]]


Example 1: a priori approach

Solve $(A+\Delta A)(x+\Delta x)=b$.



In [2]:
import numpy as np
Ap = [[.2161,.1441], [1.2969,.8648]]
Bp=[[0.1440],[0.8642]]
dAp=[[.0001,0], [0,0]]
A = np.array(Ap)
dA = np.array(dAp)
print(A+dA)
B = np.array(Bp)
print(B)
Ainv=np.linalg.inv(A+dA)
x=np.matmul(Ainv,B)
print("solution: x + $\Delta$ x=", x)


[[0.2162 0.1441]
 [1.2969 0.8648]]
[[0.144 ]
 [0.8642]]
solution: x + $\Delta$ x= [[-2.31294090e-04]
 [ 9.99653059e-01]]


A small error in one element of A, leads to big error in $x$.

$x+\Delta x=  \left[\begin{array}{c}-2.312...\times10^{-4} \\\\ 9.99....\times10^{-1} \end{array} \right]   $

Example 2

$Ax = b$ for the case where:

$(A|b) =\begin{bmatrix}  
  1 & 1.000 &| 2.000 \\
  1 & 1.001 &| 2.001 \end{bmatrix}$

  Exact solution $x=\begin{bmatrix}  1   \\ 1  \end{bmatrix}$


Make a small perturbation:

  $\Delta A =\begin{bmatrix}  
  0 & 0 \\
  0 & .001 \end{bmatrix} $

In [3]:
#Example 2 Gezerlis Ch. 4
import numpy as np
Ap = [[1,1], [1,1.001]]
Bp=[[2],[2.001]]
dAp=[[.000,0], [0,0.001]]
A = np.array(Ap)
print(A)
dA = np.array(dAp)
print(A+dA)
B = np.array(Bp)
Ainv=np.linalg.inv(A)
Adainv=np.linalg.inv(A+dA)
x=np.matmul(Ainv,B)
xdA=np.matmul(Adainv,B)
print(x)
print(xdA)

[[1.    1.   ]
 [1.    1.001]]
[[1.    1.   ]
 [1.    1.002]]
[[1.]
 [1.]]
[[1.5]
 [0.5]]


Solution: $x+\Delta x=\begin{bmatrix}  1.5   \\ .5  \end{bmatrix}$

Make a small perturbation $\Delta b= \begin{bmatrix}  0   \\ .001  \end{bmatrix}$

In [4]:
#Example 2b Gezerlis Ch. 4
import numpy as np
Ap = [[1,1], [1,1.001]]
Bp=[[2],[2.001]]
dBp=[[0], [0.001]]
A = np.array(Ap)
print(A)
dB = np.array(Bp)+np.array(dBp)
Ainv=np.linalg.inv(A)
x=np.matmul(Ainv,B)
xdb=np.matmul(Ainv,dB)
print(x)
print(xdb)

[[1.    1.   ]
 [1.    1.001]]
[[1.]
 [1.]]
[[0.]
 [2.]]


Example 3:

$Ax = b$ for the case where:

$(A|b) =\begin{bmatrix}  
  2 & 1 &| 2 \\
  1 & 2 &| 7 \end{bmatrix}$

Exact solution $x=\begin{bmatrix}  -1   \\ 4  \end{bmatrix}$


 Make a small perturbation:

  $\Delta A =\begin{bmatrix}  
  0 & 0 \\
  0.01 & 0 \end{bmatrix} $ or  $\Delta b=\begin{bmatrix}  .01   \\ 0  \end{bmatrix}$

In [5]:
#Example 3 Gezerlis Ch. 4
import numpy as np
Ap = [[2,1], [1,2]]
Bp=[[2],[7]]
dAp=[[.000,0.0], [0.01,0.0]]
dBp=[[0.01], [0.0]]
A = np.array(Ap)
print(A)
dA = np.array(dAp)
print(A+dA)
B = np.array(Bp)
dB=np.array(dBp)+np.array(Bp)
print(B)
Ainv=np.linalg.inv(A)
Adainv=np.linalg.inv(A+dA)
x=np.matmul(Ainv,B)
xdA=np.matmul(Adainv,B)
xdB=np.matmul(Ainv,dB)
print(x)
print(xdA)
print(xdB)

[[2 1]
 [1 2]]
[[2.   1.  ]
 [1.01 2.  ]]
[[2]
 [7]]
[[-1.]
 [ 4.]]
[[-1.00334448]
 [ 4.00668896]]
[[-0.99333333]
 [ 3.99666667]]


Exact solution $x= \begin{bmatrix}  -1 \\ 4\end{bmatrix}$

$\tilde{x}_{\Delta A}=\begin{bmatrix}  -1.0003344   \\ 4.006688  \end{bmatrix}$

$\tilde{x}_{\Delta b}=\begin{bmatrix}  -0.9933   \\ 3.9966  \end{bmatrix}$

Analysis

Determinant of each example


In [6]:
import numpy as np
from numpy import linalg as LA
A1 = np.array([[.2161,.1441], [1.2969,.8648]])
A2 = np.array([[1,1], [1,1.001]])
A3 = np.array([[2,1], [1,2]])
A4 =np.array([[.2,.1], [.1,.2]])
print(LA.det(A1), LA.norm(A1))
print(LA.det(A2), LA.norm(A2))
print(LA.det(A3), LA.norm(A3))
print(LA.det(A4), LA.norm(A4))

-9.999999984453023e-09 1.5802824652573981
0.00099999999999989 2.000500187453128
2.9999999999999996 3.1622776601683795
0.03000000000000001 0.31622776601683794




From these examples , we can draw a conclusion:

A should be non-singular i.e., non-zero determinant. 𝐴 Matrices with determinant close to zero are close to being singular and therefore are sensitive to very small perturbations.

What does small mean?

Example 4: What if I multiply each element of A of example 3 by 0.1 and b by 0.1 too?


In [7]:
#Example 4 Gezerlis Ch. 4
import numpy as np
from numpy import linalg as LA
Ap = [[.2,.1], [.1,.2]]
Bp=[[.2],[.7]]
dAp=[[.000,0.0], [0.01,0.0]]
dBp=[[0.01], [0.0]]
A = np.array(Ap)
print(A)
dA = np.array(dAp)
print(A+dA)
B = np.array(Bp)
dB=np.array(dBp)+np.array(Bp)
print(B)
Ainv=np.linalg.inv(A)
Adainv=np.linalg.inv(A+dA)
x=np.matmul(Ainv,B)
xdA=np.matmul(Adainv,B)
xdB=np.matmul(Ainv,dB)
print(x)
print(xdA)
print(xdB)
print(LA.det(A))
LA.norm(A)

[[0.2 0.1]
 [0.1 0.2]]
[[0.2  0.1 ]
 [0.11 0.2 ]]
[[0.2]
 [0.7]]
[[-1.]
 [ 4.]]
[[-1.03448276]
 [ 4.06896552]]
[[-0.93333333]
 [ 3.96666667]]
0.03000000000000001


0.31622776601683794

Results from example 3 and 4 are identical. Although |𝐴| for ex. 4 is much smaller than that of ex. 3. But in example 4 det was small and so were matrix elements, that means its the ratio which matters, and ratio of ex. 4 and ex. 3 were identical.

That means not just determinant but also norm matters.

Ex. 1 and 2 the ratio |𝐴|/||𝐴|| was very small

Ex. 3 : |𝐴|/||𝐴||∼𝑂(1) Ex. 4: |𝐴|/||𝐴||∼𝑂(.1)

In [8]:
#Example 5 multiply A , b of example 3 with 10^-10
import numpy as np
Ap = [[2*10**(-10),1*10**(-10)], [1*10**(-10),2*10**(-10)]]
Bp=[[2*10**(-10)],[7*10**(-10)]]
dAp=[[.000,0.0], [0.01*10**(-10),0.0]]
dBp=[[0.01*10**(-10)], [0.0]]
A = np.array(Ap)
print(A)
dA = np.array(dAp)
print(A+dA)
B = np.array(Bp)
dB=np.array(dBp)+np.array(Bp)
print(B)
Ainv=np.linalg.inv(A)
Adainv=np.linalg.inv(A+dA)
x=np.matmul(Ainv,B)
xdA=np.matmul(Adainv,B)
xdB=np.matmul(Ainv,dB)
xdX=np.matmul(Adainv,dB)
print(x)
print(xdA)
print(xdB)
print(xdX)
print(LA.det(A), LA.norm(A))

[[2.e-10 1.e-10]
 [1.e-10 2.e-10]]
[[2.00e-10 1.00e-10]
 [1.01e-10 2.00e-10]]
[[2.e-10]
 [7.e-10]]
[[-1.]
 [ 4.]]
[[-1.00334448]
 [ 4.00668896]]
[[-0.99333333]
 [ 3.99666667]]
[[-0.99665552]
 [ 4.00331104]]
3.000000000000012e-20 3.1622776601683795e-10


Even if $|A| << ||A||$ the solutions are insensitive to perturbations for ex. 5 as long we change both A and b.

Example 1: $|A| \sim 10^{-8}$, $||A||_{E}=1.58$, $|A| /||A||_{E}  << 1$ holds, therefore solution were prone to errors.


Example 2: $|A| \sim 0.001$, $||A||_{E}=2.0$, $|A| /||A||_{E}  << 1$ holds, therefore solution were prone to errors.

Example 3: $|A| = 3$, $||A||_{E}\sim 3.16$, $|A| /||A||_{E}  << 1$ does not hold, solution almost independent of errors.

Example 4: $|A| = .03$, $||A||_{E}\sim 0.32$, $|A| /||A||_{E}  << 1$ does not really hold, solution independent of errors.

Example 5: $|A| = 10^{-20}$, $||A||_{E}\sim 10^{-10}$, $|A| /||A||_{E}  << 1$ does hold, but perturbation/error creep has marginal impact, again solution is almost independent of errors.


Conclusion Det/Norm ratio $<< 1$ is not a good criterion.

Example 6: Upper triangular matrix $A$ with diagonal entries $2$ and upper diagonal entries $-2$. $b$ single column vector with alternate entries of $1,-1$. Add perturbation $\Delta A=0.01$ to row 7, col 0 element. Check how solutions are affected?


In [9]:
import numpy as np
A6= np.zeros((8, 8))
print("The upper triangular matrix: ")
for i in range(8):
   for j in range(8):
      if(i > j):
         A6[i][j] = 0
         print(A6[i][j],end=" ")
      elif(i==j):
        A6[i][j]=2
        print(A6[i][i],end=" ")
      else:
        A6[i][j]=-2
        print(A6[i][j],end=" ")
   print(" ")
print(np.linalg.det(A6))
frob_norm = np.linalg.norm(A6,ord='fro')
print(f"{frob_norm = :.2f}")
Bp=[[1],[-1],[1],[-1],[1],[-1],[1],[-1]]
B = np.array(Bp)
Ainv=np.linalg.inv(A6)
x=np.matmul(Ainv,B)
print(x)
da=np.zeros((8, 8))
da[7][0]=-0.01
Ad=A6+da
print(Ad)
print(np.linalg.det(Ad))
Adinv=np.linalg.inv(Ad)
xd=np.matmul(Adinv,B)
print(xd)

The upper triangular matrix: 
2.0 -2.0 -2.0 -2.0 -2.0 -2.0 -2.0 -2.0  
0.0 2.0 -2.0 -2.0 -2.0 -2.0 -2.0 -2.0  
0.0 0.0 2.0 -2.0 -2.0 -2.0 -2.0 -2.0  
0.0 0.0 0.0 2.0 -2.0 -2.0 -2.0 -2.0  
0.0 0.0 0.0 0.0 2.0 -2.0 -2.0 -2.0  
0.0 0.0 0.0 0.0 0.0 2.0 -2.0 -2.0  
0.0 0.0 0.0 0.0 0.0 0.0 2.0 -2.0  
0.0 0.0 0.0 0.0 0.0 0.0 0.0 2.0  
255.99999999999994
frob_norm = 12.00
[[-21. ]
 [-11. ]
 [ -5. ]
 [ -3. ]
 [ -1. ]
 [ -1. ]
 [  0. ]
 [ -0.5]]
[[ 2.   -2.   -2.   -2.   -2.   -2.   -2.   -2.  ]
 [ 0.    2.   -2.   -2.   -2.   -2.   -2.   -2.  ]
 [ 0.    0.    2.   -2.   -2.   -2.   -2.   -2.  ]
 [ 0.    0.    0.    2.   -2.   -2.   -2.   -2.  ]
 [ 0.    0.    0.    0.    2.   -2.   -2.   -2.  ]
 [ 0.    0.    0.    0.    0.    2.   -2.   -2.  ]
 [ 0.    0.    0.    0.    0.    0.    2.   -2.  ]
 [-0.01  0.    0.    0.    0.    0.    0.    2.  ]]
174.07999999999993
[[-30.88235294]
 [-15.94117647]
 [ -7.47058824]
 [ -4.23529412]
 [ -1.61764706]
 [ -1.30882353]
 [ -0.15441176]
 [ -0.65441176]]


Condition Number for previous examples

In [10]:
import numpy as np
A1 = np.array([[.2161,.1441], [1.2969,.8648]])
A2 = np.array([[1,1], [1,1.001]])
A3 = np.array([[2,1], [1,2]])
A4 = np.array([[.2,.1], [.1,.2]])
A5 = 10e-10*np.array([[2,1], [1,2]])
A6= np.zeros((8, 8))
print("The upper triangular matrix: ")
for i in range(8):
   for j in range(8):
      if(i > j):
         A6[i][j] = 0
         print(A6[i][j],end=" ")
      elif(i==j):
        A6[i][j]=2
        print(A6[i][i],end=" ")
      else:
        A6[i][j]=-2
        print(A6[i][j],end=" ")
   print(" ")
print(np.linalg.norm(np.linalg.inv(A1))*np.linalg.norm(A1))
print(np.linalg.norm(np.linalg.inv(A2))*np.linalg.norm(A2))
print(np.linalg.norm(np.linalg.inv(A3))*np.linalg.norm(A3))
print(np.linalg.norm(np.linalg.inv(A4))*np.linalg.norm(A4))
print(np.linalg.norm(np.linalg.inv(A5))*np.linalg.norm(A5))
print(np.linalg.norm(np.linalg.inv(A6))*np.linalg.norm(A6))

The upper triangular matrix: 
2.0 -2.0 -2.0 -2.0 -2.0 -2.0 -2.0 -2.0  
0.0 2.0 -2.0 -2.0 -2.0 -2.0 -2.0 -2.0  
0.0 0.0 2.0 -2.0 -2.0 -2.0 -2.0 -2.0  
0.0 0.0 0.0 2.0 -2.0 -2.0 -2.0 -2.0  
0.0 0.0 0.0 0.0 2.0 -2.0 -2.0 -2.0  
0.0 0.0 0.0 0.0 0.0 2.0 -2.0 -2.0  
0.0 0.0 0.0 0.0 0.0 0.0 2.0 -2.0  
0.0 0.0 0.0 0.0 0.0 0.0 0.0 2.0  
249729267.38825405
4002.0010000004404
3.333333333333334
3.3333333333333326
3.333333333333333
512.183560845133


Calculating Eigenvalues of a Matrix

Example 8

$A =\begin{bmatrix}  
  4 & 3 & 2 & 1 \\
  3 & 3 & 2 & 1 \\
  0 & 2 & 2 & 1 \\
  0 & 0 &1 & 1 \end{bmatrix}$

  If I add a perturbation to element $A_{0,3}=0.1$, what is the effect on the eigenvalues?
  

In [11]:
import numpy as np
A8 = np.array([[4,3,2,1], [3,3,2,1],[0,2,2,1],[0,0,1,1]])
w,v=np.linalg.eig(A8)
print('E-value:', w)
print('E-vector', v)
A8norm = np.linalg.norm(A8)
A8inv=np.linalg.inv(A8)
A8invnorm=np.linalg.norm(A8inv)
print(A8norm*A8invnorm)
A8d = np.array([[4,3,2,1+0.01], [3,3,2,1],[0,2,2,1],[0,0,1,1]])
w,v=np.linalg.eig(A8d)
print('E-value:', w)
print('E-vector', v)

E-value: [7.31274213 2.06663092 0.13674761 0.48387934]
E-vector [[ 0.73319229 -0.50597951 -0.03552739  0.08981592]
 [ 0.63293    -0.26114648  0.22427524 -0.09580043]
 [ 0.24559146  0.599716   -0.63638388 -0.45466534]
 [ 0.03890408  0.56225259  0.73719331  0.88092838]]
126.74383614203887
E-value: [7.3129755  2.06287308 0.12477715 0.49937427]
E-vector [[ 0.73322121 -0.50489056 -0.04175813  0.09211235]
 [ 0.63290497 -0.2628705   0.23432976 -0.11006428]
 [ 0.24557035  0.59881226 -0.63967137 -0.44302653]
 [ 0.0388993   0.56339018  0.73086686  0.88494557]]


Example 9

$A =\begin{bmatrix}  
  4 & 4 & 0 & 0 \\
  0 & 3 & 4 & 0 \\
  0 & 0 & 2 & 4 \\
  0 & 0 &0 & 1 \end{bmatrix}$

  If I add a perturbation to element $A_{3,0}=0.005$, what is the effect on the eigenvalues?

In [12]:
import numpy as np
A9 = np.array([[4,4,0,0], [0,3,4,0],[0,0,2,4],[0,0,0,1]])
w,v=np.linalg.eig(A9)
print('E-value:', w)
print('E-vector', v)
A9norm = np.linalg.norm(A9)
A9inv=np.linalg.inv(A9)
A9invnorm=np.linalg.norm(A9inv)
print(A9norm*A9invnorm)
A9d = np.array([[4,4,0,0], [0,3,4,0],[0,0,2,4],[0.005,0,0,1]])
w,v=np.linalg.eig(A9d)
print('E-value:', w)
print('E-vector', v)

E-value: [4. 3. 2. 1.]
E-vector [[ 1.         -0.9701425   0.88888889 -0.76429148]
 [ 0.          0.24253563 -0.44444444  0.57321861]
 [ 0.          0.          0.11111111 -0.28660931]
 [ 0.          0.          0.          0.07165233]]
40.129477943277564
E-value: [4.04884232 0.95115768 2.18205744 2.81794256]
E-vector [[ 0.99991899  0.75725717 -0.90713923 -0.95891902]
 [ 0.01220959 -0.57718943  0.41228175  0.28337434]
 [ 0.00320148  0.29564253 -0.0843057  -0.0128976 ]
 [ 0.00163983 -0.0775206  -0.00383712 -0.00263737]]


This is much better conditioned, at least better than ex. 8, but changes are large compared to perturbation.

CONDITION NUMBER FOR EIGEN VALUES Example 8 (again)

In [13]:
import numpy as np
from scipy.linalg import eig
A8 = np.array([[4,3,2,1], [3,3,2,1],[0,2,2,1],[0,0,1,1]])
w,v=np.linalg.eig(A8)
print('E-value:', w)
print('E-vector', v)
w, vl, vr = eig(A8, left=True)
idx = np.argsort(w)[::-1]
eigenValues = w[idx]
LeigenVectors =vl[:,idx]
eigenVectors = vr[:,idx]
print(eigenValues, eigenVectors)
vr3=eigenVectors[:,3]
vl3=LeigenVectors[:,3]
print(vr3)
vlt3=vl3.T
K3=1/abs(np.matmul(vlt3,vr3))
print(K3)
A8norm = np.linalg.norm(A8)
A8inv=np.linalg.inv(A8)
A8invnorm=np.linalg.norm(A8inv)
print(A8norm*A8invnorm)
A8d = np.array([[4,3,2,1+0.01], [3,3,2,1],[0,2,2,1],[0,0,1,1]])
w,v=np.linalg.eig(A8d)
print('E-value:', w)
print('E-vector', v)

E-value: [7.31274213 2.06663092 0.13674761 0.48387934]
E-vector [[ 0.73319229 -0.50597951 -0.03552739  0.08981592]
 [ 0.63293    -0.26114648  0.22427524 -0.09580043]
 [ 0.24559146  0.599716   -0.63638388 -0.45466534]
 [ 0.03890408  0.56225259  0.73719331  0.88092838]]
[7.31274213+0.j 2.06663092+0.j 0.48387934+0.j 0.13674761+0.j] [[ 0.73319229 -0.50597951  0.08981592 -0.03552739]
 [ 0.63293    -0.26114648 -0.09580043  0.22427524]
 [ 0.24559146  0.599716   -0.45466534 -0.63638388]
 [ 0.03890408  0.56225259  0.88092838  0.73719331]]
[-0.03552739  0.22427524 -0.63638388  0.73719331]
2.8230996335945195
126.74383614203887
E-value: [7.3129755  2.06287308 0.12477715 0.49937427]
E-vector [[ 0.73322121 -0.50489056 -0.04175813  0.09211235]
 [ 0.63290497 -0.2628705   0.23432976 -0.11006428]
 [ 0.24557035  0.59881226 -0.63967137 -0.44302653]
 [ 0.0388993   0.56339018  0.73086686  0.88494557]]


Back to example 9

In [14]:
import numpy as np
from scipy.linalg import eig
A9 = np.array([[4,4,0,0], [0,3,4,0],[0,0,2,4],[0,0,0,1]])
w,v=np.linalg.eig(A9)
print('E-value:', w)
print('E-vector', v)
w, vl, vr = eig(A9, left=True)
idx = np.argsort(w)[::-1]
eigenValues = w[idx]
LeigenVectors =vl[:,idx]
eigenVectors = vr[:,idx]
print(eigenValues, eigenVectors)
vr2=eigenVectors[:,2]
vl2=LeigenVectors[:,2]
print(vr2)
vlt2=vl2.T
K2=1/abs(np.matmul(vlt2,vr2))
print(K2)
A9norm = np.linalg.norm(A9)
A9inv=np.linalg.inv(A9)
A9invnorm=np.linalg.norm(A9inv)
print(A9norm*A9invnorm)
A9d = np.array([[4,4,0,0], [0,3,4,0],[0,0,2,4],[0.005,0,0,1]])
w,v=np.linalg.eig(A9d)
idx = np.argsort(w)[::-1]
eigenValues = w[idx]
eigenVectors = v[:,idx]
print('E-value:', eigenValues)
print('E-vector', eigenVectors[:,0])
print(eigenVectors[:,1])
print(eigenVectors[:,2])
print(eigenVectors[:,3])

E-value: [4. 3. 2. 1.]
E-vector [[ 1.         -0.9701425   0.88888889 -0.76429148]
 [ 0.          0.24253563 -0.44444444  0.57321861]
 [ 0.          0.          0.11111111 -0.28660931]
 [ 0.          0.          0.          0.07165233]]
[4.+0.j 3.+0.j 2.+0.j 1.+0.j] [[ 1.         -0.9701425   0.88888889 -0.76429148]
 [ 0.          0.24253563 -0.44444444  0.57321861]
 [ 0.          0.          0.11111111 -0.28660931]
 [ 0.          0.          0.          0.07165233]]
[ 0.88888889 -0.44444444  0.11111111  0.        ]
37.107950630558946
40.129477943277564
E-value: [4.04884232 2.81794256 2.18205744 0.95115768]
E-vector [0.99991899 0.01220959 0.00320148 0.00163983]
[-0.95891902  0.28337434 -0.0128976  -0.00263737]
[-0.90713923  0.41228175 -0.0843057  -0.00383712]
[ 0.75725717 -0.57718943  0.29564253 -0.0775206 ]


An eigen value condition number differs from condition number for linear equations. An eigenvalue condition number close to 1 means its very well conditioned. While an eigenvalue condition much greater than 1 implies ill conditioned.



Sensitivity to eigenvectors

Example 3 (Again)


In [15]:
import numpy as np
from scipy.linalg import eig
A3 = np.array([[2,1],[1,2]])
w, vl, vr = eig(A3, left=True)
idx = np.argsort(w)[::-1]
eigenValues = w[idx]
LeigenVectors =vl[:,idx]
eigenVectors = vr[:,idx]
print(eigenValues, eigenVectors[:,0],eigenVectors[:,1])
A3d = np.array([[2,1],[1+.01,2]])
w, vl, vr = eig(A3d, left=True)
idx = np.argsort(w)[::-1]
eigenValues = w[idx]
LeigenVectors =vl[:,idx]
eigenVectors = vr[:,idx]
print(eigenValues, eigenVectors[:,0],eigenVectors[:,1])

[3.+0.j 1.+0.j] [0.70710678 0.70710678] [-0.70710678  0.70710678]
[3.00498756+0.j 0.99501244+0.j] [0.70534562 0.70886357] [-0.70534562  0.70886357]


This is example of well conditioned, because error in eigenvectors is actually much less than the magnitude of the perturbation.

In [16]:
import numpy as np
from scipy.linalg import eig
A9 = np.array([[4,4,0,0],[0,3,4,0],[0,0,2,4],[0,0,0,1]])
w, vl, vr = eig(A9, left=True)
idx = np.argsort(w)[::-1]
eigenValues = w[idx]
LeigenVectors =vl[:,idx]
eigenVectors = vr[:,idx]
print(eigenValues)
print(eigenVectors[:,0])
print(eigenVectors[:,1])
print(eigenVectors[:,2])
print(eigenVectors[:,3])
A9d = np.array([[4,4,0,0],[0,3,4,0],[0,0,2,4],[0+.005,0,0,1]])
w, vl, vr = eig(A9d, left=True)
idx = np.argsort(w)[::-1]
eigenValues = w[idx]
LeigenVectors =vl[:,idx]
eigenVectors = vr[:,idx]
print(eigenValues),
print(eigenVectors[:,0])
print(eigenVectors[:,1])
print(eigenVectors[:,2])
print(eigenVectors[:,3])

[4.+0.j 3.+0.j 2.+0.j 1.+0.j]
[1. 0. 0. 0.]
[-0.9701425   0.24253563  0.          0.        ]
[ 0.88888889 -0.44444444  0.11111111  0.        ]
[-0.76429148  0.57321861 -0.28660931  0.07165233]
[4.04884232+0.j 2.81794256+0.j 2.18205744+0.j 0.95115768+0.j]
[0.99991899 0.01220959 0.00320148 0.00163983]
[-0.95891902  0.28337434 -0.0128976  -0.00263737]
[-0.90713923  0.41228175 -0.0843057  -0.00383712]
[ 0.75725717 -0.57718943  0.29564253 -0.0775206 ]


Example 10

$A =\begin{bmatrix}  
  1.01 & 0.01  \\
  0 & 0.99  \end{bmatrix}$

  If I add a perturbation to element $A_{0,1}=0.005$, what is the effect on the eigenvalues and eigenvectors?

In [17]:
import numpy as np
from scipy.linalg import eig
A10 = np.array([[1.01,.01],[0,.99]])
w, vl, vr = eig(A10, left=True)
idx = np.argsort(w)[::-1]
eigenValues = w[idx]
LeigenVectors =vl[:,idx]
eigenVectors = vr[:,idx]
vr0=eigenVectors[:,0]
vl0=LeigenVectors[:,0]
print(vr0)
vlt0=vl0.T
print(vlt0)
K0=1/abs(np.matmul(vlt0,vr0))
print(K0)
vr1=eigenVectors[:,1]
vl1=LeigenVectors[:,1]
print(vr1)
vlt1=vl1.T
print(vlt1)
K1=1/abs(np.matmul(vlt1,vr1))
print(K1)
print(eigenValues, eigenVectors[:,0],eigenVectors[:,1])
A10d = np.array([[1.01,.01+.005],[0,.99]])
w, vl, vr = eig(A10d, left=True)
idx = np.argsort(w)[::-1]
eigenValues = w[idx]
LeigenVectors =vl[:,idx]
eigenVectors = vr[:,idx]
print(eigenValues, eigenVectors[:,0],eigenVectors[:,1])

[1. 0.]
[0.89442719 0.4472136 ]
1.1180339887498947
[-0.4472136   0.89442719]
[0. 1.]
1.1180339887498947
[1.01+0.j 0.99+0.j] [1. 0.] [-0.4472136   0.89442719]
[1.01+0.j 0.99+0.j] [1. 0.] [-0.6  0.8]


Example 11

$A =\begin{bmatrix}  
  1 & 2 & 3  \\
  0 & 4 & 5 \\
  0 & 0 & 4.001  \end{bmatrix}$

  If I add a perturbation to element $A_{2,0}=0.005$, what is the effect on the eigenvalues and eigenvectors?

In [18]:
import numpy as np
from scipy.linalg import eig
A11 = np.array([[1,2,3],[0,4,5],[0,0,4.001]])
w, vl, vr = eig(A11, left=True)
idx = np.argsort(w)[::-1]
eigenValues = w[idx]
LeigenVectors =vl[:,idx]
eigenVectors = vr[:,idx]
vr0=eigenVectors[:,0]
vl0=LeigenVectors[:,0]
print(vr0)
vlt0=vl0.T
print(vlt0)
K0=1/abs(np.matmul(vlt0,vr0))
print(K0)
vr1=eigenVectors[:,1]
vl1=LeigenVectors[:,1]
print(vr1)
vlt1=vl1.T
print(vlt1)
K1=1/abs(np.matmul(vlt1,vr1))
print(K1)
print(eigenValues, eigenVectors[:,0],eigenVectors[:,1],eigenVectors[:,2])
A11d = np.array([[1,2,3],[0,4,5],[0.005,0,4.001]])
w, vl, vr = eig(A11d, left=True)
idx = np.argsort(w)[::-1]
eigenValues = w[idx]
LeigenVectors =vl[:,idx]
eigenVectors = vr[:,idx]
print(eigenValues, eigenVectors[:,0],eigenVectors[:,1],eigenVectors[:,2])

[5.54687392e-01 8.32058814e-01 1.66411763e-04]
[0. 0. 1.]
6009.190596870348
[0.5547002  0.83205029 0.        ]
[ 0.00000000e+00  1.99999996e-04 -9.99999980e-01]
6009.25224595635
[4.001+0.j 4.   +0.j 1.   +0.j] [5.54687392e-01 8.32058814e-01 1.66411763e-04] [0.5547002  0.83205029 0.        ] [1. 0. 0.]
[4.12933336+0.j 3.87111014+0.j 1.0005565 +0.j] [-0.55298164 -0.83291484 -0.02154473] [-0.55629299 -0.83071027  0.02141403] [ 0.99999475  0.00277787 -0.00166641]


Solving $Ax=b$ for lower and upper triangular matrices

In [ ]:
import numpy as np
def forsub(L,bs):
  n = bs.size
  xs = np.zeros(n)
  for i in range(n):
    xs[i] = (bs[i] - L[i,:i]@xs[:i])/L[i,i]
  return xs
def backsub(U,bs):
  n = bs.size
  xs = np.zeros(n)
  for i in reversed(range(n)):
    xs[i] = (bs[i] - U[i,i+1:]@xs[i+1:])/U[i,i]
  return xs
def testcreate(n,val):
  A = np.arange(val,val+n*n).reshape(n,n)
  A = np.sqrt(A)
  bs = (A[0,:])**2.1
  return A, bs
def testsolve(f,A,bs):
  xs = f(A,bs); print(xs)
  xs = np.linalg.solve(A,bs); print(xs)

A, bs = testcreate(4,21)
print(A)
print(bs)
L = np.tril(A)
print(L)
testsolve(forsub,L,bs)
print(" ")
U = np.triu(A)
print(U)
testsolve(backsub,U,bs)

[[4.58257569 4.69041576 4.79583152 4.89897949]
 [5.         5.09901951 5.19615242 5.29150262]
 [5.38516481 5.47722558 5.56776436 5.65685425]
 [5.74456265 5.83095189 5.91607978 6.        ]]
[24.45289367 25.67697243 26.90383729 28.13337297]
[[4.58257569 0.         0.         0.        ]
 [5.         5.09901951 0.         0.        ]
 [5.38516481 5.47722558 5.56776436 0.        ]
 [5.74456265 5.83095189 5.91607978 6.        ]]
[ 5.33605887 -0.19676761 -0.13541854 -0.09524368]
[ 5.33605887 -0.19676761 -0.13541854 -0.09524368]
 
[[4.58257569 4.69041576 4.79583152 4.89897949]
 [0.         5.09901951 5.19615242 5.29150262]
 [0.         0.         5.56776436 5.65685425]
 [0.         0.         0.         6.        ]]
[0.14941285 0.10032435 0.06814924 4.6888955 ]
[0.14941285 0.10032435 0.06814924 4.6888955 ]


Gaussian elimination


In [ ]:
import numpy as np
def gauelim(inA,inbs):
  A = np.copy(inA)
  bs = np.copy(inbs)
  n = bs.size
  for j in range(n-1):
    for i in range(j+1,n):
      coeff = A[i,j]/A[j,j]
      A[i,j:] -= coeff*A[j,j:]
      bs[i] -= coeff*bs[j]
  xs = backsub(A,bs)
  return xs

def forsub(L,bs):
  n = bs.size
  xs = np.zeros(n)
  for i in range(n):
    xs[i] = (bs[i] - L[i,:i]@xs[:i])/L[i,i]
  return xs
def backsub(U,bs):
  n = bs.size
  xs = np.zeros(n)
  for i in reversed(range(n)):
    xs[i] = (bs[i] - U[i,i+1:]@xs[i+1:])/U[i,i]
  return xs
def testcreate(n,val):
  A = np.arange(val,val+n*n).reshape(n,n)
  A = np.sqrt(A)
  bs = (A[0,:])**2.1
  return A, bs
def testsolve(f,A,bs):
  xs = f(A,bs); print(xs)
  xs = np.linalg.solve(A,bs); print(xs)
#if _name == '_main_':
A, bs = testcreate(4,21)
print(A)
print(bs)
testsolve(gauelim,A,bs)

[[4.58257569 4.69041576 4.79583152 4.89897949]
 [5.         5.09901951 5.19615242 5.29150262]
 [5.38516481 5.47722558 5.56776436 5.65685425]
 [5.74456265 5.83095189 5.91607978 6.        ]]
[24.45289367 25.67697243 26.90383729 28.13337297]
[ 17118.9554009  -55069.99934969  58822.07580723 -20866.39246612]
[ 17118.95550576 -55069.99968225  58822.07615809 -20866.39258928]


LU Decomposition

In [ ]:
import numpy as np
def ludec(A):
  n = A.shape[0]
  U = np.copy(A)
  L = np.identity(n)
  for j in range(n-1):
    for i in range(j+1,n):
      coeff = U[i,j]/U[j,j]
      U[i,j:] -= coeff*U[j,j:]
      L[i,j] = coeff
  return L, U
def lusolve(A,bs):
  L, U = ludec(A)
  ys = forsub(L,bs)
  xs = backsub(U,ys)
  return xs
def forsub(L,bs):
  n = bs.size
  xs = np.zeros(n)
  for i in range(n):
    xs[i] = (bs[i] - L[i,:i]@xs[:i])/L[i,i]
  return xs
def backsub(U,bs):
  n = bs.size
  xs = np.zeros(n)
  for i in reversed(range(n)):
    xs[i] = (bs[i] - U[i,i+1:]@xs[i+1:])/U[i,i]
  return xs
def testcreate(n,val):
  A = np.arange(val,val+n*n).reshape(n,n)
  A = np.sqrt(A)
  bs = (A[0,:])**2.1
  return A, bs
def testsolve(f,A,bs):
  xs = f(A,bs); print(xs)
  xs = np.linalg.solve(A,bs); print(xs)
A, bs = testcreate(4,21)
testsolve(lusolve,A,bs)

[ 17118.9554009  -55069.99934969  58822.07580723 -20866.39246612]
[ 17118.95550576 -55069.99968225  58822.07615809 -20866.39258928]


Gaussian elimnation doesn't work: \\
Example 1:
$(A|b)= \begin{bmatrix}  
  0 & -1 &| 1 \\
  1 & 1 &| 2 \end{bmatrix}$

In [ ]:
import numpy as np
def gauelim(inA,inbs):
  A = np.copy(inA)
  bs = np.copy(inbs)
  n = bs.size
  for j in range(n-1):
    for i in range(j+1,n):
      coeff = A[i,j]/A[j,j]
      A[i,j:] -= coeff*A[j,j:]
      bs[i] -= coeff*bs[j]
  xs = backsub(A,bs)
  return xs

def forsub(L,bs):
  n = bs.size
  xs = np.zeros(n)
  for i in range(n):
    xs[i] = (bs[i] - L[i,:i]@xs[:i])/L[i,i]
  return xs
def backsub(U,bs):
  n = bs.size
  xs = np.zeros(n)
  for i in reversed(range(n)):
    xs[i] = (bs[i] - U[i,i+1:]@xs[i+1:])/U[i,i]
  return xs
Ap=[[0.,-1.], [1.,1.]]
bs=[1.,2.]
A = np.matrix(Ap)
print(A)
print(bs)
def testsolve(f,A,bs):
  xs = f(A,bs); print(xs)
  xs = np.linalg.solve(A,bs); print(xs)
testsolve(gauelim,A,bs)

[[ 0. -1.]
 [ 1.  1.]]
[1.0, 2.0]
[nan nan]
[ 3. -1.]


<ipython-input-2-e1312c4eee02>:8: RuntimeWarning: divide by zero encountered in scalar divide
  coeff = A[i,j]/A[j,j]
<ipython-input-2-e1312c4eee02>:9: RuntimeWarning: invalid value encountered in multiply
  A[i,j:] -= coeff*A[j,j:]
<ipython-input-2-e1312c4eee02>:24: RuntimeWarning: invalid value encountered in scalar divide
  xs[i] = (bs[i] - U[i,i+1:]@xs[i+1:])/U[i,i]


Example 2:
$(A|b)= \begin{bmatrix}  
  2 & 1 &1 & | 8 \\
  2 & 1 &-4  & |  -2 \\
  1 & 2 & 1  &| 2 \end{bmatrix}$

In [ ]:
import numpy as np
def gauelim(inA,inbs):
  A = np.copy(inA)
  bs = np.copy(inbs)
  n = bs.size
  for j in range(n-1):
    for i in range(j+1,n):
      coeff = A[i,j]/A[j,j]
      A[i,j:] -= coeff*A[j,j:]
      bs[i] -= coeff*bs[j]
  xs = backsub(A,bs)
  return xs

def forsub(L,bs):
  n = bs.size
  xs = np.zeros(n)
  for i in range(n):
    xs[i] = (bs[i] - L[i,:i]@xs[:i])/L[i,i]
  return xs
def backsub(U,bs):
  n = bs.size
  xs = np.zeros(n)
  for i in reversed(range(n)):
    xs[i] = (bs[i] - U[i,i+1:]@xs[i+1:])/U[i,i]
  return xs
Ap=[[2.,1.,1.], [2.,1.,-4.],[1.,2.,1.]]
bs=[8.,-2.,2.]
A = np.matrix(Ap)
print(A)
print(bs)
def testsolve(f,A,bs):
  xs = f(A,bs); print(xs)
  xs = np.linalg.solve(A,bs); print(xs)
testsolve(gauelim,A,bs)

[[ 2.  1.  1.]
 [ 2.  1. -4.]
 [ 1.  2.  1.]]
[8.0, -2.0, 2.0]
[nan nan nan]
[ 4. -2.  2.]


<ipython-input-3-40b4fdcefc11>:8: RuntimeWarning: divide by zero encountered in scalar divide
  coeff = A[i,j]/A[j,j]
<ipython-input-3-40b4fdcefc11>:9: RuntimeWarning: invalid value encountered in multiply
  A[i,j:] -= coeff*A[j,j:]
<ipython-input-3-40b4fdcefc11>:24: RuntimeWarning: invalid value encountered in scalar divide
  xs[i] = (bs[i] - U[i,i+1:]@xs[i+1:])/U[i,i]


Gaussian elimnation doesn't work: \\
Example 3:
$(A|b)= \begin{bmatrix}  
  10^{-20} & -1 &   | 1 \\
  1 & 1 &   | 2 \end{bmatrix}$

In [ ]:
import numpy as np
def gauelim(inA,inbs):
  A = np.copy(inA)
  bs = np.copy(inbs)
  n = bs.size
  for j in range(n-1):
    for i in range(j+1,n):
      coeff = A[i,j]/A[j,j]
      A[i,j:] -= coeff*A[j,j:]
      bs[i] -= coeff*bs[j]
  xs = backsub(A,bs)
  return xs
def backsub(U,bs):
  n = bs.size
  xs = np.zeros(n)
  for i in reversed(range(n)):
    xs[i] = (bs[i] - U[i,i+1:]@xs[i+1:])/U[i,i]
  return xs
Ap=[[10.**(-20),-1.], [1.,1.]]
bs=[1.,2.]
A = np.matrix(Ap)
print(A)
print(bs)
def testsolve(f,A,bs):
  xs = f(A,bs); print(xs)
  xs = np.linalg.solve(A,bs); print(xs)
testsolve(gauelim,A,bs)

[[ 1.e-20 -1.e+00]
 [ 1.e+00  1.e+00]]
[1.0, 2.0]
[ 0. -1.]
[ 3. -1.]


PARTIAL PIVOTING

Example 1:
$(A|b)= \begin{bmatrix}  
  0 & -1 &| 1 \\
  1 & 1 &| 2 \end{bmatrix}$

In [ ]:
import numpy as np
def gauelim_pivot(inA,inbs):
  A = np.copy(inA)
  bs = np.copy(inbs)
  n = bs.size
  for j in range(n-1):
    k = np.argmax(np.abs(A[j:,j])) + j
#fIND INDEX OF MAXIMUM ENTRY IN COLUMN j
    if k != j:
      A[j,:], A[k,:] = A[k,:], A[j,:].copy()
#INTERCHANGE ROWS j and k
      bs[j], bs[k] = bs[k], bs[j]
    for i in range(j+1,n):
      coeff = A[i,j]/A[j,j]
      A[i,j:] -= coeff*A[j,j:]
      bs[i] -= coeff*bs[j]
  xs = backsub(A,bs)
  return xs
def backsub(U,bs):
  n = bs.size
  xs = np.zeros(n)
  for i in reversed(range(n)):
    xs[i] = (bs[i] - U[i,i+1:]@xs[i+1:])/U[i,i]
  return xs
Ap=[[0.,-1.], [1.,1.]]
bs=[1.,2.]
A = np.matrix(Ap)
print(A)
print(bs)
def testsolve(f,A,bs):
  xs = f(A,bs); print(xs)
  xs = np.linalg.solve(A,bs); print(xs)
testsolve(gauelim_pivot,A,bs)

[[ 0. -1.]
 [ 1.  1.]]
[1.0, 2.0]
[ 3. -1.]
[ 3. -1.]


Partial Pivot
Example 2:
$(A|b)= \begin{bmatrix}  
  2 & 1 & 1 &| 8 \\
  2 & 1 & -4&|  -2 \\
  1 & 2 & 1 &| 2 \end{bmatrix}$

In [ ]:
import numpy as np
def gauelim_pivot(inA,inbs):
  A = np.copy(inA)
  bs = np.copy(inbs)
  n = bs.size
  for j in range(n-1):
    k = np.argmax(np.abs(A[j:,j])) + j
    if k != j:
      A[j,:], A[k,:] = A[k,:], A[j,:].copy()
      bs[j], bs[k] = bs[k], bs[j]
    for i in range(j+1,n):
      coeff = A[i,j]/A[j,j]
      A[i,j:] -= coeff*A[j,j:]
      bs[i] -= coeff*bs[j]
  xs = backsub(A,bs)
  return xs
def backsub(U,bs):
  n = bs.size
  xs = np.zeros(n)
  for i in reversed(range(n)):
    xs[i] = (bs[i] - U[i,i+1:]@xs[i+1:])/U[i,i]
  return xs
Ap=[[2.,1.,1.], [2.,1.,-4.],[1.,2.,1.]]
bs=[8.,-2.,2]
A = np.array(Ap)
def testsolve(f,A,bs):
  xs = f(A,bs); print(xs)
  xs = np.linalg.solve(A,bs); print(xs)
testsolve(gauelim_pivot,A,bs)

[ 4. -2.  2.]
[ 4. -2.  2.]


Partial Pivot
Example 3:
$(A|b)= \begin{bmatrix}  
  10^{-20} & -1 &| 1 \\
  1 & 1 &| 2 \end{bmatrix}$

In [ ]:
import numpy as np
def gauelim_pivot(inA,inbs):
  A = np.copy(inA)
  bs = np.copy(inbs)
  n = bs.size
  for j in range(n-1):
    k = np.argmax(np.abs(A[j:,j])) + j
    if k != j:
      A[j,:], A[k,:] = A[k,:], A[j,:].copy()
      bs[j], bs[k] = bs[k], bs[j]
    for i in range(j+1,n):
      coeff = A[i,j]/A[j,j]
      A[i,j:] -= coeff*A[j,j:]
      bs[i] -= coeff*bs[j]
  xs = backsub(A,bs)
  return xs
def backsub(U,bs):
  n = bs.size
  xs = np.zeros(n)
  for i in reversed(range(n)):
    xs[i] = (bs[i] - U[i,i+1:]@xs[i+1:])/U[i,i]
  return xs
Ap=[[10**(-20),-1.], [1.,1.]]
bs=[1.,2]
A = np.matrix(Ap)
def testsolve(f,A,bs):
  xs = f(A,bs); print(xs)
  xs = np.linalg.solve(A,bs); print(xs)
testsolve(gauelim_pivot,A,bs)

[ 3. -1.]
[ 3. -1.]


SCALED PARTIAL




Define for each row: a scale factor $s_{i}= max_{0\le j \le n-1}  |A_{ij}|$.


$s_i$ is the largest element in row $i$.


Instead of determining $k$ as corresponding to largest matrix element in column $j$ (at or below $A_{jj}$), as done in Partial Pivot, we determine $k$ by finding matrix element with largest relative magnitude, i.e., the largest in comparison to other elements in its row.


Thus we search for smallest integer $k$ that satisfies:

$\frac{|A_{kj}|}{s_{k}} = \max_{j\le m\le n-1} \frac{|A_{mj}|}{s_{m}}  $.

Then we interchange rows $j$ and $k$.

In [ ]:
#scaled partial pivot
import numpy as np
def gauelim_pivot_scale(inA,inbs):
    A = np.copy(inA)
    bs = np.copy(inbs)
    n = bs.size
    scales = np.max(np.abs(A),axis=1)
#In a 2-dimensional NumPy array, the axes are the directions along the rows and columns.
#AXIS 0 IS THE DIRECTION ALONG THE ROWS, AXIS 1 IS THE DIRECTION ALONG THE COLUMNS

    for j in range(n-1):
        k = np.argmax(np.abs(A[j:,j])/scales[j:]) + j
        scales[j], scales[k] = scales[k], scales[j]
        bs[j], bs[k] = bs[k], bs[j]
        A[j,:], A[k,:] = A[k,:], A[j,:].copy()

        for i in range(j+1,n):
            coeff = A[i,j]/A[j,j]
            A[i,j:] -= coeff*A[j,j:]
            bs[i] -= coeff*bs[j]

    xs = backsub(A,bs)
    return xs
def backsub(U,bs):
  n = bs.size
  xs = np.zeros(n)
  for i in reversed(range(n)):
    xs[i] = (bs[i] - U[i,i+1:]@xs[i+1:])/U[i,i]
  return xs
Ap=[[10**(-20),-1.], [10**(-20),10**(-20)]]
bs=[1,2.*10**(-20)]
A = np.matrix(Ap)
print(A)
print(bs)
def testsolve(f,A,bs):
  xs = f(A,bs); print(xs)
  xs = np.linalg.solve(A,bs); print(xs)
testsolve(gauelim_pivot_scale,A,bs)

[[ 1.e-20 -1.e+00]
 [ 1.e-20  1.e-20]]
[1, 2e-20]
[ 3. -1.]
[ 0. -1.]


Jacobi Iterative Method


$x_i^{k}=(b_{i} - \sum_{j=0}^{i-1} A_{ij}x_{j}^{k-1}- \sum_{j=i+1}^{n-1} A_{ij}x_{j}^{k-1})\frac{1}{A_{ii}}, i=0, 1, ..n-1$


convergence criteria:

$\sum_{i=0}^{n-1} |\frac{x_{i}^{k}-x_{i}^{k-1}}{x_{i}^k}| \le \epsilon$

In [ ]:
import numpy as np
def termcrit(xolds,xnews):
  errs = np.abs((xnews - xolds)/xnews)
  return np.sum(errs)
def jacobi(A,bs,kmax=50,tol=1.e-6):
  n = bs.size
  xnews = np.zeros(n)
  for k in range(1,kmax):
    xs = np.copy(xnews)
    for i in range(n):
      slt = A[i,:i]@xs[:i]
      sgt = A[i,i+1:]@xs[i+1:]
      xnews[i] = (bs[i] - slt - sgt)/A[i,i]
    err = termcrit(xs, xnews)
    print(k, xnews, err)
    if err < tol:
      break
  else:
    xnews = None
  return xnews
def testcreate(n,val):
  A = np.arange(val,val+n*n).reshape(n,n)
  A = np.sqrt(A)
  bs = (A[0,:])**2.1
  return A, bs
def testsolve(f,A,bs):
  xs = f(A,bs); print(xs)
  xs = np.linalg.solve(A,bs); print(xs)
n = 4; val = 21;
A, bs = testcreate(n,val)
A += val*np.identity(n)
testsolve(jacobi,A,bs)

1 [0.95584174 0.98382901 1.01264965 1.04197678] 4.0
2 [0.38609117 0.38784058 0.39421864 0.40425572] 6.158643474272702
3 [0.73341768 0.74941422 0.76835846 0.78969405] 1.9310672038734116
4 [0.52317706 0.53023862 0.54134654 0.55573114] 1.6555533665830167
5 [0.65072149 0.6631481  0.67896251 0.6975372 ] 0.8024076458188099
6 [0.57339977 0.58256412 0.59551555 0.61154376] 0.5539167512173229
7 [0.6202852  0.63142596 0.64611142 0.66368227] 0.30983792417040845
8 [0.59185737 0.60179949 0.61543315 0.63206832] 0.19712628238915642
9 [0.60909428 0.61976313 0.63403445 0.65123688] 0.11605599760545227
10 [0.59864294 0.60887114 0.62275581 0.63961427] 0.07162932168524415
11 [0.60497996 0.61547534 0.62959445 0.64666147] 0.04296480237007709
12 [0.6011376  0.61147098 0.62544794 0.6423885 ] 0.026221914973545257
13 [0.60346736 0.61389897 0.62796212 0.64497935] 0.015836323357587702
14 [0.60205475 0.61242679 0.62643768 0.64340843] 0.009625240606544867
15 [0.60291127 0.61331942 0.627362   0.64436094] 0.00582762295